pkg/basicstation/cups/update_info.go

```diff
@@ -229,68 +236,81 @@ func (s *Server) UpdateInfo(c echo.Context) error {
 		gtw.Attributes = make(map[string]string)
 	}
 
-	if gtw.LBSLNSSecret == nil {
-		return errLNSCredentials.WithAttributes("gateway_uid", uid)
-	}
-
-	if s.requireExplicitEnable || gtw.Attributes[cupsAttribute] != "" {
+	if s.requireExplicitEnable && gtw.Attributes[cupsAttribute] != "" {
 		if cups, _ := strconv.ParseBool(gtw.Attributes[cupsAttribute]); !cups {
 			return errCUPSNotEnabled.WithAttributes("gateway_uid", uid)
 		}
 	}
 
 	res := UpdateInfoResponse{}
+	if s.allowCUPSURIUpdate && gtw.TargetCUPSURI != "" && gtw.TargetCUPSURI != req.CUPSURI {
+		if gtw.TargetCUPSKey.Value == nil {
+			return errTargetCUPSCredentials.New()
+		}
+		logger := logger.WithField("cups_uri", gtw.TargetCUPSURI)
+		logger.Debug("Configure CUPS")
+		res.CUPSURI = gtw.TargetCUPSURI
 
-	if gtw.Attributes[cupsURIAttribute] == "" {
-		gtw.Attributes[cupsURIAttribute] = req.CUPSURI
-	}
-	if s.allowCUPSURIUpdate && gtw.Attributes[cupsURIAttribute] != req.CUPSURI {
-		res.CUPSURI = gtw.Attributes[cupsURIAttribute]
-	}
-
-	if credentials := gtw.Attributes[cupsCredentialsAttribute]; credentials != "" {
-		logger.WithField("cups_uri", gtw.Attributes[cupsURIAttribute]).Debug("Getting trusted certificate for CUPS...")
-		cupsTrust, err := s.getTrust(gtw.Attributes[cupsURIAttribute])
+		cupsTrust, err := s.getTrust(gtw.TargetCUPSURI)
+		if err != nil {
+			return errServerTrust.WithCause(err).WithAttributes("address", gtw.TargetCUPSURI)
+		}
+		cupsCredentials, err := TokenCredentials(cupsTrust, string(gtw.TargetCUPSKey.Value))
 		if err != nil {
 			return err
 		}
-		cupsCredentials, err := TokenCredentials(cupsTrust, credentials)
+		if crc32.ChecksumIEEE(cupsCredentials) != req.CUPSCredentialsCRC {
+			res.CUPSCredentials = cupsCredentials
+		}
+	} else if gtw.TargetCUPSKey != nil && gtw.TargetCUPSKey.Value != nil {
+		// Check if CUPS Key needs to be rotated.
+		cupsTrust, err := s.getTrust(req.CUPSURI)
+		if err != nil {
+			return errServerTrust.WithCause(err).WithAttributes("address", req.CUPSURI)
+		}
+		cupsCredentials, err := TokenCredentials(cupsTrust, string(gtw.TargetCUPSKey.Value))
 		if err != nil {
 			return err
 		}
 		if crc32.ChecksumIEEE(cupsCredentials) != req.CUPSCredentialsCRC {
+			logger.Debug("Update CUPS Credentials")
 			res.CUPSCredentials = cupsCredentials
+			res.CUPSURI = req.CUPSURI
+		}
+	} else {
```

Comment:
I think this entire if/elif/else block can be structured a bit better:

```go
if s.allowCUPSURIUpdate && gtw.TargetCUPSURI != "" {
  // we are allowed to and may want to configure another CUPS
  if gtw.TargetCUPSKey.Value == nil || gtw.TargetCUPSKey.Value == nil {
    // but for that we do need credentials
    return errTargetCUPSCredentials.New()
  }
  if gtw.TargetCUPSURI != req.CUPSURI {
    // we need to update the CUPS URI
    res.CUPSURI = gtw.TargetCUPSURI
  }
  cupsTrust, err := s.getTrust(gtw.TargetCUPSURI)
  // ...
  cupsCredentials, err := TokenCredentials(cupsTrust, string(gtw.TargetCUPSKey.Value))
  // ...
  if crc32.ChecksumIEEE(cupsCredentials) != req.CUPSCredentialsCRC {
    res.CUPSURI = req.CUPSURI
  }
} else {
  // we are not allowed to configure another CUPS, or we don't have one
  // so we configure LNS
```

Reply:
hmm. That doesn't account for the case where we need to only do a credentials update. `gtw.TargetCUPSURI != ""` would need to be set update only the credentials. This is why I avoided this scheme.